drivers/serial/uart_mcux_iuart.c

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright (c) 2019, Manivannan Sadhasivam <mani@kernel.org>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_kinetis_iuart
+
+#include <device.h>
+#include <drivers/uart.h>
+#include <drivers/clock_control.h>
+#include <errno.h>
+#include <fsl_uart.h>
+#include <soc.h>
+
+struct mcux_iuart_config {
+	UART_Type *base;
+	char *clock_name;
+	clock_control_subsys_t clock_subsys;
+	u32_t baud_rate;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	void (*irq_config_func)(struct device *dev);
+#endif
+};
+
+struct mcux_iuart_data {
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t callback;
+	void *cb_data;
+#endif
+};
+
+#define DEV_CFG(dev)						\
+	((const struct mcux_iuart_config * const)(dev)->config_info)
+
+static int mcux_iuart_poll_in(struct device *dev, unsigned char *c)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	int ret = -1;
+
+	if (UART_GetStatusFlag(config->base, kUART_RxDataReadyFlag)) {
+		*c = UART_ReadByte(config->base);
+		ret = 0;
+	}
+
+	return ret;
+}
+
+static void mcux_iuart_poll_out(struct device *dev, unsigned char c)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+
+	while (!(UART_GetStatusFlag(config->base, kUART_TxReadyFlag))) {
+	}
+
+	UART_WriteByte(config->base, c);
+}
+
+static int mcux_iuart_err_check(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	int err = 0;
+
+	if (UART_GetStatusFlag(config->base, kUART_RxOverrunFlag)) {
+		err |= UART_ERROR_OVERRUN;
+		UART_ClearStatusFlag(config->base, kUART_RxOverrunFlag);
+	}
+
+	if (UART_GetStatusFlag(config->base, kUART_ParityErrorFlag)) {
+		err |= UART_ERROR_PARITY;
+		UART_ClearStatusFlag(config->base, kUART_ParityErrorFlag);
+	}
+
+	if (UART_GetStatusFlag(config->base, kUART_FrameErrorFlag)) {
+		err |= UART_ERROR_FRAMING;
+		UART_ClearStatusFlag(config->base, kUART_FrameErrorFlag);
+	}
+
+	return err;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static int mcux_iuart_fifo_fill(struct device *dev, const u8_t *tx_data,
+			       int len)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	u8_t num_tx = 0U;
+
+	while ((len - num_tx > 0) &&
+	       (UART_GetStatusFlag(config->base, kUART_TxEmptyFlag))) {
+
+		UART_WriteByte(config->base, tx_data[num_tx++]);
+	}
+
+	return num_tx;
+}
+
+static int mcux_iuart_fifo_read(struct device *dev, u8_t *rx_data,
+			       const int len)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	u8_t num_rx = 0U;
+
+	while ((len - num_rx > 0) &&
+	       (UART_GetStatusFlag(config->base, kUART_RxDataReadyFlag))) {
+
+		rx_data[num_rx++] = UART_ReadByte(config->base);
+	}
+
+	return num_rx;
+}
+
+static void mcux_iuart_irq_tx_enable(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	u32_t mask = kUART_TxEmptyEnable;
```

Comment:
We could directly pass `kUART_TxEmptyEnable` to the interrupts enable function instead of creating a local variable.. 

The compiler will do this for us anyways, but still better to remove `mask`..